Fix unit tests with Python 3.12

### Summary

They currently fail.

### Issue Type

Bug Report

### Component Name

unit tests

### Ansible Version

devel

### Community.general Version

main

### Configuration

_No response_

### OS / Environment

_No response_

### Steps to Reproduce

`ansible-test units --docker --python 3.12`

### Expected Results

Pass

### Actual Results

```
FAILED tests/unit/plugins/modules/test_dnsimple.py::TestDNSimple::test_account_token - AttributeError: 'TestDNSimple' object has no attribute 'assertEquals'. Did ...
FAILED tests/unit/plugins/modules/test_dnsimple.py::TestDNSimple::test_user_token_single_account - AttributeError: 'TestDNSimple' object has no attribute 'assertEquals'. Did ...
FAILED tests/unit/plugins/modules/test_pagerduty.py::PagerDutyTest::test_absent_maintenance_compatibility_header - AttributeError: 'PagerDutyTest' object has no attribute 'assertDictContains...
FAILED tests/unit/plugins/modules/test_pkgin.py::TestPkginQueryPackage::test_package_found_but_not_installed - AttributeError: 'TestPkginQueryPackage' object has no attribute 'assertEqua...
FAILED tests/unit/plugins/modules/test_pagerduty.py::PagerDutyTest::test_create_maintenance_compatibility_header - AttributeError: 'PagerDutyTest' object has no attribute 'assertDictContains...
FAILED tests/unit/plugins/modules/test_pkgin.py::TestPkginQueryPackage::test_package_found_outdated - AttributeError: 'TestPkginQueryPackage' object has no attribute 'assertEqua...
FAILED tests/unit/plugins/modules/test_pkgin.py::TestPkginQueryPackage::test_package_not_found - AttributeError: 'TestPkginQueryPackage' object has no attribute 'assertEqua...
FAILED tests/unit/plugins/modules/test_pagerduty.py::PagerDutyTest::test_create_maintenance_from_header - AttributeError: 'PagerDutyTest' object has no attribute 'assertDictContains...
FAILED tests/unit/plugins/modules/test_pkgin.py::TestPkginQueryPackage::test_package_with_version_found_outdated - AttributeError: 'TestPkginQueryPackage' object has no attribute 'assertEqua...
FAILED tests/unit/plugins/modules/test_pagerduty.py::PagerDutyTest::test_ongoing_maintenance_windos_compatibility_header - AttributeError: 'PagerDutyTest' object has no attribute 'assertDictContains...
FAILED tests/unit/plugins/modules/test_pkgin.py::TestPkginQueryPackage::test_package_with_version_is_present - AttributeError: 'TestPkginQueryPackage' object has no attribute 'assertEqua...
FAILED tests/unit/plugins/modules/test_pagerduty_alert.py::PagerDutyAlertsTest::test_compatibility_header - AttributeError: 'PagerDutyAlertsTest' object has no attribute 'assertDictCo...
FAILED tests/unit/plugins/modules/test_pkgin.py::TestPkginQueryPackage::test_package_without_version_is_present - AttributeError: 'TestPkginQueryPackage' object has no attribute 'assertEqua...
FAILED tests/unit/plugins/modules/test_pkgin.py::TestPkginQueryPackage::test_with_parseable_flag_not_supported_package_is_present - AttributeError: 'TestPkginQueryPackage' object has no attribute 'assertEqua...
FAILED tests/unit/plugins/modules/test_pkgin.py::TestPkginQueryPackage::test_with_parseable_flag_supported_package_is_present - AttributeError: 'TestPkginQueryPackage' object has no attribute 'assertEqua...
FAILED tests/unit/plugins/modules/test_slack.py::TestSlackModule::test_edit_message - AttributeError: 'TestSlackModule' object has no attribute 'assertEquals'. D...
FAILED tests/unit/plugins/modules/test_slack.py::TestSlackModule::test_ts_in_message_does_not_cause_edit - AttributeError: 'TestSlackModule' object has no attribute 'assertEquals'. D...
FAILED tests/unit/plugins/modules/test_jenkins_build_info.py::TestJenkinsBuildInfo::test_module_get_build_info_get_last_build - AttributeError: 'TestJenkinsBuildInfo' object has no attribute 'assertEqual...
FAILED tests/unit/plugins/modules/test_jenkins_build_info.py::TestJenkinsBuildInfo::test_module_get_build_info_if_build_does_not_exist - AttributeError: 'TestJenkinsBuildInfo' object has no attribute 'assertEqual...
FAILED tests/unit/plugins/modules/test_jenkins_build_info.py::TestJenkinsBuildInfo::test_module_get_build_info_if_job_does_not_exist - AttributeError: 'TestJenkinsBuildInfo' object has no attribute 'assertEqual...
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

--- a/.azure-pipelines/azure-pipelines.yml
+++ b/.azure-pipelines/azure-pipelines.yml
@@ -127,7 +127,7 @@ stages:
             - test: 3.9
             - test: '3.10'
             - test: '3.11'
-            # - test: '3.12'  TODO
+            - test: '3.12'
   - stage: Units_2_16
     displayName: Units 2.16
     dependsOn: []

--- a/tests/unit/plugins/callback/test_loganalytics.py
+++ b/tests/unit/plugins/callback/test_loganalytics.py
@@ -12,6 +12,7 @@ from ansible_collections.community.general.plugins.callback.loganalytics import 
 from datetime import datetime
 
 import json
+import sys
 
 
 class TestAzureLogAnalytics(unittest.TestCase):
@@ -26,6 +27,10 @@ class TestAzureLogAnalytics(unittest.TestCase):
         self.task_fields = {'args': {}}
         self.mock_host = Mock('MockHost')
         self.mock_host.name = 'myhost'
+
+        # Add backward compatibility
+        if sys.version_info < (3, 2):
+            self.assertRegex = self.assertRegexpMatches
 
     @patch('ansible_collections.community.general.plugins.callback.loganalytics.datetime')
     @patch('ansible_collections.community.general.plugins.callback.loganalytics.open_url')
@@ -62,6 +67,5 @@ class TestAzureLogAnalytics(unittest.TestCase):
         args, kwargs = open_url_mock.call_args
         headers = kwargs['headers']
 
-        (self.assertRegex if getattr(self, 'assertRegex') else self.assertRegexpMatches)(
-            headers['Authorization'], r'^SharedKey 01234567-0123-0123-0123-01234567890a:.*=$')
+        self.assertRegex(headers['Authorization'], r'^SharedKey 01234567-0123-0123-0123-01234567890a:.*=$')
         self.assertEqual(headers['Log-Type'], 'ansible_playbook')

--- a/tests/unit/plugins/callback/test_loganalytics.py
+++ b/tests/unit/plugins/callback/test_loganalytics.py
@@ -62,5 +62,6 @@ class TestAzureLogAnalytics(unittest.TestCase):
         args, kwargs = open_url_mock.call_args
         headers = kwargs['headers']
 
-        self.assertRegexMatches(headers['Authorization'], r'^SharedKey 01234567-0123-0123-0123-01234567890a:.*=$')
+        (self.assertRegex if getattr(self, 'assertRegex') else self.assertRegexpMatches)(
+            headers['Authorization'], r'^SharedKey 01234567-0123-0123-0123-01234567890a:.*=$')
         self.assertEqual(headers['Log-Type'], 'ansible_playbook')

--- a/tests/unit/plugins/callback/test_loganalytics.py
+++ b/tests/unit/plugins/callback/test_loganalytics.py
@@ -62,5 +62,5 @@ class TestAzureLogAnalytics(unittest.TestCase):
         args, kwargs = open_url_mock.call_args
         headers = kwargs['headers']
 
-        self.assertRegexpMatches(headers['Authorization'], r'^SharedKey 01234567-0123-0123-0123-01234567890a:.*=$')
+        self.assertRegexMatches(headers['Authorization'], r'^SharedKey 01234567-0123-0123-0123-01234567890a:.*=$')
         self.assertEqual(headers['Log-Type'], 'ansible_playbook')

--- a/tests/unit/plugins/module_utils/hwc/test_hwc_utils.py
+++ b/tests/unit/plugins/module_utils/hwc/test_hwc_utils.py
@@ -29,12 +29,13 @@ class HwcUtilsTestCase(unittest.TestCase):
                            {"foo.quiet.trees": 1}),
             1)
 
-        self.assertRaisesRegex(HwcModuleException,
-                               r".* key\(q\) is not exist in dict",
-                               navigate_value, value, ["foo", "q", "tree"])
+        assertRaisesRegex = self.assertRaisesRegex if getattr(self, 'assertRaisesRegex') else self.assertRaisesRegexp
+        assertRaisesRegex(HwcModuleException,
+                          r".* key\(q\) is not exist in dict",
+                          navigate_value, value, ["foo", "q", "tree"])
 
-        self.assertRaisesRegex(HwcModuleException,
-                               r".* the index is out of list",
-                               navigate_value, value,
-                               ["foo", "quiet", "trees"],
-                               {"foo.quiet.trees": 2})
+        assertRaisesRegex(HwcModuleException,
+                          r".* the index is out of list",
+                          navigate_value, value,
+                          ["foo", "quiet", "trees"],
+                          {"foo.quiet.trees": 2})

--- a/tests/unit/plugins/module_utils/hwc/test_hwc_utils.py
+++ b/tests/unit/plugins/module_utils/hwc/test_hwc_utils.py
@@ -29,12 +29,12 @@ class HwcUtilsTestCase(unittest.TestCase):
                            {"foo.quiet.trees": 1}),
             1)
 
-        self.assertRaisesRegexp(HwcModuleException,
-                                r".* key\(q\) is not exist in dict",
-                                navigate_value, value, ["foo", "q", "tree"])
+        self.assertRaisesRegex(HwcModuleException,
+                               r".* key\(q\) is not exist in dict",
+                               navigate_value, value, ["foo", "q", "tree"])
 
-        self.assertRaisesRegexp(HwcModuleException,
-                                r".* the index is out of list",
-                                navigate_value, value,
-                                ["foo", "quiet", "trees"],
-                                {"foo.quiet.trees": 2})
+        self.assertRaisesRegex(HwcModuleException,
+                               r".* the index is out of list",
+                               navigate_value, value,
+                               ["foo", "quiet", "trees"],
+                               {"foo.quiet.trees": 2})

--- a/tests/unit/plugins/module_utils/hwc/test_hwc_utils.py
+++ b/tests/unit/plugins/module_utils/hwc/test_hwc_utils.py
@@ -6,11 +6,20 @@
 from __future__ import (absolute_import, division, print_function)
 __metaclass__ = type
 
+import sys
+
 from ansible_collections.community.general.tests.unit.compat import unittest
 from ansible_collections.community.general.plugins.module_utils.hwc_utils import (HwcModuleException, navigate_value)
 
 
 class HwcUtilsTestCase(unittest.TestCase):
+    def setUp(self):
+        super(HwcUtilsTestCase, self).setUp()
+
+        # Add backward compatibility
+        if sys.version_info < (3, 0):
+            self.assertRaisesRegex = self.assertRaisesRegexp
+
     def test_navigate_value(self):
         value = {
             'foo': {
@@ -29,13 +38,12 @@ class HwcUtilsTestCase(unittest.TestCase):
                            {"foo.quiet.trees": 1}),
             1)
 
-        assertRaisesRegex = self.assertRaisesRegex if getattr(self, 'assertRaisesRegex') else self.assertRaisesRegexp
-        assertRaisesRegex(HwcModuleException,
-                          r".* key\(q\) is not exist in dict",
-                          navigate_value, value, ["foo", "q", "tree"])
+        self.assertRaisesRegex(HwcModuleException,
+                               r".* key\(q\) is not exist in dict",
+                               navigate_value, value, ["foo", "q", "tree"])
 
-        assertRaisesRegex(HwcModuleException,
-                          r".* the index is out of list",
-                          navigate_value, value,
-                          ["foo", "quiet", "trees"],
-                          {"foo.quiet.trees": 2})
+        self.assertRaisesRegex(HwcModuleException,
+                               r".* the index is out of list",
+                               navigate_value, value,
+                               ["foo", "quiet", "trees"],
+                               {"foo.quiet.trees": 2})

--- a/tests/unit/plugins/modules/test_dnsimple.py
+++ b/tests/unit/plugins/modules/test_dnsimple.py
@@ -45,7 +45,7 @@ class TestDNSimple(ModuleTestCase):
     def test_account_token(self, mock_whoami):
         mock_whoami.return_value.data.account = 42
         ds = self.module.DNSimpleV2('fake', 'fake', True, self.module)
-        self.assertEquals(ds.account, 42)
+        self.assertEqual(ds.account, 42)
 
     @patch('dnsimple.service.Accounts.list_accounts')
     @patch('dnsimple.service.Identity.whoami')
@@ -61,4 +61,4 @@ class TestDNSimple(ModuleTestCase):
         mock_accounts.return_value.data = [42]
         mock_whoami.return_value.data.account = None
         ds = self.module.DNSimpleV2('fake', 'fake', True, self.module)
-        self.assertEquals(ds.account, 42)
+        self.assertEqual(ds.account, 42)

--- a/tests/unit/plugins/modules/test_jenkins_build_info.py
+++ b/tests/unit/plugins/modules/test_jenkins_build_info.py
@@ -140,7 +140,7 @@ class TestJenkinsBuildInfo(unittest.TestCase):
 
         self.assertFalse(return_json.exception.args[0]['changed'])
         self.assertTrue(return_json.exception.args[0]['failed'])
-        self.assertEquals("ABSENT", return_json.exception.args[0]['build_info']['result'])
+        self.assertEqual("ABSENT", return_json.exception.args[0]['build_info']['result'])
 
     @patch('ansible_collections.community.general.plugins.modules.jenkins_build_info.test_dependencies')
     @patch('ansible_collections.community.general.plugins.modules.jenkins_build_info.JenkinsBuildInfo.get_jenkins_connection')
@@ -157,7 +157,7 @@ class TestJenkinsBuildInfo(unittest.TestCase):
             jenkins_build_info.main()
 
         self.assertFalse(return_json.exception.args[0]['changed'])
-        self.assertEquals("SUCCESS", return_json.exception.args[0]['build_info']['result'])
+        self.assertEqual("SUCCESS", return_json.exception.args[0]['build_info']['result'])
 
     @patch('ansible_collections.community.general.plugins.modules.jenkins_build_info.test_dependencies')
     @patch('ansible_collections.community.general.plugins.modules.jenkins_build_info.JenkinsBuildInfo.get_jenkins_connection')
@@ -177,4 +177,4 @@ class TestJenkinsBuildInfo(unittest.TestCase):
 
         self.assertFalse(return_json.exception.args[0]['changed'])
         self.assertTrue(return_json.exception.args[0]['failed'])
-        self.assertEquals("ABSENT", return_json.exception.args[0]['build_info']['result'])
+        self.assertEqual("ABSENT", return_json.exception.args[0]['build_info']['result'])

--- a/tests/unit/plugins/modules/test_pagerduty.py
+++ b/tests/unit/plugins/modules/test_pagerduty.py
@@ -20,9 +20,9 @@ class PagerDutyTest(unittest.TestCase):
         return object(), {'status': 200}
 
     def _assert_ongoing_window_with_v1_compatible_header(self, module, url, headers, data=None, method=None):
-        self.assertDictContainsSubset(
-            {'Accept': 'application/vnd.pagerduty+json;version=2'},
-            headers,
+        self.assertEqual(
+            'application/vnd.pagerduty+json;version=2',
+            headers.get('Accept'),
             'Accept:application/vnd.pagerduty+json;version=2 HTTP header not found'
         )
         return object(), {'status': 200}
@@ -36,17 +36,17 @@ class PagerDutyTest(unittest.TestCase):
         return object(), {'status': 201}
 
     def _assert_create_a_maintenance_window_from_header(self, module, url, headers, data=None, method=None):
-        self.assertDictContainsSubset(
-            {'From': 'requester_id'},
-            headers,
+        self.assertEqual(
+            'requester_id',
+            headers.get('From'),
             'From:requester_id HTTP header not found'
         )
         return object(), {'status': 201}
 
     def _assert_create_window_with_v1_compatible_header(self, module, url, headers, data=None, method=None):
-        self.assertDictContainsSubset(
-            {'Accept': 'application/vnd.pagerduty+json;version=2'},
-            headers,
+        self.assertEqual(
+            'application/vnd.pagerduty+json;version=2',
+            headers.get('Accept'),
             'Accept:application/vnd.pagerduty+json;version=2 HTTP header not found'
         )
         return object(), {'status': 201}
@@ -89,9 +89,9 @@ class PagerDutyTest(unittest.TestCase):
         return object(), {'status': 204}
 
     def _assert_absent_window_with_v1_compatible_header(self, module, url, headers, method=None):
-        self.assertDictContainsSubset(
-            {'Accept': 'application/vnd.pagerduty+json;version=2'},
-            headers,
+        self.assertEqual(
+            'application/vnd.pagerduty+json;version=2',
+            headers.get('Accept'),
             'Accept:application/vnd.pagerduty+json;version=2 HTTP header not found'
         )
         return object(), {'status': 204}

--- a/tests/unit/plugins/modules/test_pagerduty_alert.py
+++ b/tests/unit/plugins/modules/test_pagerduty_alert.py
@@ -22,9 +22,9 @@ class PagerDutyAlertsTest(unittest.TestCase):
         return Response(), {'status': 200}
 
     def _assert_compatibility_header(self, module, url, method, headers):
-        self.assertDictContainsSubset(
-            {'Accept': 'application/vnd.pagerduty+json;version=2'},
-            headers,
+        self.assertEqual(
+            'application/vnd.pagerduty+json;version=2',
+            headers.get('Accept'),
             'Accept:application/vnd.pagerduty+json;version=2 HTTP header not found'
         )
         return Response(), {'status': 200}

--- a/tests/unit/plugins/modules/test_pkgin.py
+++ b/tests/unit/plugins/modules/test_pkgin.py
@@ -30,7 +30,7 @@ class TestPkginQueryPackage(unittest.TestCase):
         command_result = pkgin.query_package(mock_module, package)
 
         # then
-        self.assertEquals(command_result, pkgin.PackageState.PRESENT)
+        self.assertEqual(command_result, pkgin.PackageState.PRESENT)
 
     @mock.patch('ansible_collections.community.general.plugins.modules.pkgin.AnsibleModule')
     def test_package_with_version_is_present(self, mock_module):
@@ -46,7 +46,7 @@ class TestPkginQueryPackage(unittest.TestCase):
         command_result = pkgin.query_package(mock_module, package)
 
         # then
-        self.assertEquals(command_result, pkgin.PackageState.PRESENT)
+        self.assertEqual(command_result, pkgin.PackageState.PRESENT)
 
     @mock.patch('ansible_collections.community.general.plugins.modules.pkgin.AnsibleModule')
     def test_package_found_but_not_installed(self, mock_module):
@@ -62,7 +62,7 @@ class TestPkginQueryPackage(unittest.TestCase):
         command_result = pkgin.query_package(mock_module, package)
 
         # then
-        self.assertEquals(command_result, pkgin.PackageState.NOT_INSTALLED)
+        self.assertEqual(command_result, pkgin.PackageState.NOT_INSTALLED)
 
     @mock.patch('ansible_collections.community.general.plugins.modules.pkgin.AnsibleModule')
     def test_package_found_outdated(self, mock_module):
@@ -78,7 +78,7 @@ class TestPkginQueryPackage(unittest.TestCase):
         command_result = pkgin.query_package(mock_module, package)
 
         # then
-        self.assertEquals(command_result, pkgin.PackageState.OUTDATED)
+        self.assertEqual(command_result, pkgin.PackageState.OUTDATED)
 
     @mock.patch('ansible_collections.community.general.plugins.modules.pkgin.AnsibleModule')
     def test_package_with_version_found_outdated(self, mock_module):
@@ -94,7 +94,7 @@ class TestPkginQueryPackage(unittest.TestCase):
         command_result = pkgin.query_package(mock_module, package)
 
         # then
-        self.assertEquals(command_result, pkgin.PackageState.OUTDATED)
+        self.assertEqual(command_result, pkgin.PackageState.OUTDATED)
 
     @mock.patch('ansible_collections.community.general.plugins.modules.pkgin.AnsibleModule')
     def test_package_not_found(self, mock_module):
@@ -110,7 +110,7 @@ class TestPkginQueryPackage(unittest.TestCase):
         command_result = pkgin.query_package(mock_module, package)
 
         # then
-        self.assertEquals(command_result, pkgin.PackageState.NOT_FOUND)
+        self.assertEqual(command_result, pkgin.PackageState.NOT_FOUND)
 
     @mock.patch('ansible_collections.community.general.plugins.modules.pkgin.AnsibleModule')
     def test_with_parseable_flag_supported_package_is_present(self, mock_module):
@@ -126,7 +126,7 @@ class TestPkginQueryPackage(unittest.TestCase):
         command_result = pkgin.query_package(mock_module, package)
 
         # then
-        self.assertEquals(command_result, pkgin.PackageState.PRESENT)
+        self.assertEqual(command_result, pkgin.PackageState.PRESENT)
 
     @mock.patch('ansible_collections.community.general.plugins.modules.pkgin.AnsibleModule')
     def test_with_parseable_flag_not_supported_package_is_present(self, mock_module):
@@ -142,4 +142,4 @@ class TestPkginQueryPackage(unittest.TestCase):
         command_result = pkgin.query_package(mock_module, package)
 
         # then
-        self.assertEquals(command_result, pkgin.PackageState.PRESENT)
+        self.assertEqual(command_result, pkgin.PackageState.PRESENT)

--- a/tests/unit/plugins/modules/test_slack.py
+++ b/tests/unit/plugins/modules/test_slack.py
@@ -105,7 +105,7 @@ class TestSlackModule(ModuleTestCase):
                 self.module.main()
 
             self.assertTrue(fetch_url_mock.call_count, 1)
-            self.assertEquals(fetch_url_mock.call_args[1]['url'], "https://slack.com/api/chat.postMessage")
+            self.assertEqual(fetch_url_mock.call_args[1]['url'], "https://slack.com/api/chat.postMessage")
 
     def test_edit_message(self):
         set_module_args({
@@ -125,9 +125,9 @@ class TestSlackModule(ModuleTestCase):
                 self.module.main()
 
             self.assertTrue(fetch_url_mock.call_count, 2)
-            self.assertEquals(fetch_url_mock.call_args[1]['url'], "https://slack.com/api/chat.update")
+            self.assertEqual(fetch_url_mock.call_args[1]['url'], "https://slack.com/api/chat.update")
             call_data = json.loads(fetch_url_mock.call_args[1]['data'])
-            self.assertEquals(call_data['ts'], "12345")
+            self.assertEqual(call_data['ts'], "12345")
 
     def test_message_with_blocks(self):
         """tests sending a message with blocks"""

